drivers/w1/w1_net.c

```diff
@@ -0,0 +1,405 @@
+		return -EIO;
+	}
+	if (ret == 0) {
+		*last_discrepancy = W1_SEARCH_NO_DEVICE;
+		return 0;
+	}
+
+	ret = w1_write_byte(dev, command);
+	if (ret < 0) {
+		return -EIO;
+	}
+	next_discrepancy = W1_SEARCH_LAST_DEVICE;
+
+	for (size_t id_bit_nr = 1; id_bit_nr < 65; id_bit_nr++) {
+		bool last_id_bit;
+		bool last_complement_id_bit;
+
+		ret = w1_read_bit(dev);
+		if (ret < 0) {
+			return -EIO;
+		}
+		last_id_bit = (bool)ret;
+		ret = w1_read_bit(dev);
+		if (ret < 0) {
+			return -EIO;
+		}
+		last_complement_id_bit = (bool)ret;
+
+		if (last_id_bit && last_complement_id_bit) {
+			/*
+			 * No device participating:
+			 * We can stop following the branch.
+			 */
+			LOG_DBG("No device paricipating");
+			*last_discrepancy = W1_SEARCH_NO_DEVICE;
+			return 0;
+		} else if (last_id_bit != last_complement_id_bit) {
+			/*
+			 * All devices connected have same ROM bit value:
+			 * We can directly follow last_id_bit branch.
+			 */
+		} else {
+			/*
+			 * Discrepancy detected: bit value at id_bit_nr does
+			 * not match for all devices on the bus.
+			 */
+			if ((id_bit_nr > *last_discrepancy) ||
+			    ((id_bit_nr < *last_discrepancy) &&
+			     (*rom_inv_64 & BIT64(id_bit_nr - 1)))) {
+				/*
+				 * - id_bit_nr > last_discrepancy:
+				 *     Start always w/ branch of 1s
+				 * - id_bit_nr < last_discrepancy:
+				 *     Follow same branch as before
+				 */
+				last_id_bit = true;
+				next_discrepancy = id_bit_nr;
+			} else {
+				/*
+				 * - id_bit_nr == last_discrepancy:
+				 *     1-path already done, therefore go 0 path
+				 * - id_bit_nr < last_discrepancy:
+				 *     Follow same branch as before
+				 */
+			}
+		}
+
+		/*
+		 * Send and store the chosen bit: all not matching devices will
+		 * no longer participate in this search until they are reset.
+		 */
+		ret = w1_write_bit(dev, last_id_bit);
+		if (ret < 0) {
+			return -EIO;
+		}
+		*rom_inv_64 &= ~BIT64(id_bit_nr - 1);
+		*rom_inv_64 |= last_id_bit ? BIT64(id_bit_nr - 1) : 0;
+	}
+
+	*last_discrepancy = next_discrepancy;
+	return 0;
+}
+
+int z_impl_w1_search_bus(const struct device *dev,
+			    uint8_t command, uint8_t family,
+			    w1_search_callback_t callback_isr,
+			    void *callback_arg)
+{
+	size_t last_discrepancy = W1_SEARCH_DISCREPANCY_INIT;
+	uint64_t found_rom_inv_64 = 0;
+	struct w1_rom found_rom = {0};
+	int found_cnt = 0;
+	int ret;
+
+	w1_lock_bus(dev);
+
+	do {
+		ret = search_device(dev, command, family, &last_discrepancy,
+				    &found_rom_inv_64);
+		if (ret < 0) {
+			found_cnt = ret;
+			goto out;
+		}
+		if (last_discrepancy == W1_SEARCH_NO_DEVICE) {
+			break;
+		}
+
+		found_cnt++;
+		/*
+		 * rom is stored in found_rom_inv_64 in "inverse byte order" =>
+		 * Only big-endian targets need to swap, such that struct's
+		 * bytes are stored in big-endian byte order.
+		 */
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+		sys_memcpy_swap(&found_rom, &found_rom_inv_64, 8);
+#else
+		*(uint64_t *)&found_rom = found_rom_inv_64;
+#endif
```

Comment:
For the sake of readability, I might do it like this instead:
```suggestion
		if (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__) {
			sys_memcpy_swap(&found_rom, &found_rom_inv_64, 8);
		} else {
			*(uint64_t *)&found_rom = found_rom_inv_64;
		}
```